src/main/java/org/apache/sling/rewriter/impl/TransformerFactoryServiceTracker.java

```diff
@@ -46,7 +50,7 @@ private boolean isGlobal(final ServiceReference ref) {
     private TransformerFactoryEntry[][] cached = EMPTY_DOUBLE_ENTRY_ARRAY;
 
     /** flag for cache. */
-    private boolean cacheIsValid = true;
+    private volatile boolean cacheIsValid = true;
```

Comment:
instead of wrapping all access into ```synchronized``` blocks, wouldn't be the use of an ```AtomicBoolean``` a better way to deal with it?

Reply:
Good point, I guess the flag could be set as the last instruction in addingService/removedService and a simple compareAndSet can be used later on

Reply:
Ah, Thank you.. I'll change that

Reply:
Hi Guys,
I think, even if we use AtomicBoolean(compareAndSet) there is a possibily of error.
- Let's say a service is removed and cacheIsValid is set to false.
- Now, there is a call for getting the Global factories which enters the sync block. At this time if a service is added and sets the cache variable to false, the get method at the end will set the cacheIsValid to true and cache will not be updated with the newly added service. So, I believe we should update the cache variable only inside a sync block.
Please let me know if I am missing something.

Reply:
I fear it is even more complicated than this; I looked at how the tracker works in more detail: the tracker updates its state *after* it returns from addingService, so within that method we have no chance to correctly invalidate the cache.
The service tracker has a tracking count (getTrackingCount()) - so instead of the boolean flag, we can store the last used tracking count in the get method and compare it with the current tracking count. If they differ, the cache needs updates. The minor downside is that the cache will also be changed if there was a change in non-global transformers, but that is neglectable.

Reply:
Updated the code with usage of "getTrackingCount", removed the isCacheValid variable. Does it look fine now?